File: DSA/DivideandConquer/Mergesort.java
import java.util.Arrays;

public class Mergesort {
    static public void mergeSort(int arr[],int si,int ei){
        if(si>=ei){
            return;
        }
        int mid=si +(ei-si)/2;
        mergeSort(arr,si,mid);
        mergeSort(arr,mid+1,ei);
        merge(arr,si,mid,ei);
    }
    static public void merge(int arr[],int si,int mid,int ei){
        int temp[]=new int[ei-si+1];
        int i=si;
        int j=mid+1;
        int k=0;
        while (i<=mid && j<=ei){
            if (arr[i]<arr[j]){
                temp[k]=arr[i];
                i++;
            }
            else {
                temp[k]=arr[j];
                j++;
            }
            k++;
        }
        while(i<=mid){
            temp[k++]=arr[i++];
        }
        while (j<=ei){
            temp[k++]=arr[j++];
        }
        for (i=si,k=0;k<temp.length;i++,k++){
            arr[i]=temp[k];
        }

    }
    public static void main(String[] args) {
        int list[]={2,3,4,5,1,9};
        int start=0;
        int end=list.length-1;
        mergeSort(list,start,end);
        System.out.println(Arrays.toString(list));}
}

File: python/mergesort.py
def mergesort(x, st, en):
    if st >= en:
        return
    mid = (st + en) // 2
    mergesort(x, st, mid)
    mergesort(x, mid + 1, en)
    merge(x, st, mid, en)


def merge(x, st, mid, en):
    left = x[st:mid + 1]
    right = x[mid + 1:en + 1]
    i = 0
    j = 0
    k = st

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            x[k] = left[i]
            i += 1
        else:
            x[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        x[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        x[k] = right[j]
        j += 1
        k += 1


a = [1, 5, 3, 2, 6, 5, 7]
mergesort(a, 0, len(a) - 1)
print(a)

    